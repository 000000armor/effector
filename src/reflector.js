//@flow

import {type ID, createIDType} from './id'

const nextID = createIDType()
export class Reflector<Field, State> {
  id: ID
  seq: number
  value: $ReadOnly<Field>
  defaultValue: $ReadOnly<Field>
  getter: (parent: State) => Field
  setter: (parent: State, value: Field) => State
}

export function init<Field, State>(
  defaultValue: Field,
  getter: (parent: State) => Field,
  setter: (parent: State, value: Field) => State
): Reflector<Field, State> {
  const result = new Reflector
  result.id = nextID()
  result.seq = 0
  result.value = defaultValue
  result.defaultValue = defaultValue
  result.getter = getter
  result.setter = setter
  return result
}

export function update<Field, State>(
  parent: State,
  updater: (
    // ( (parent: State, value: Field) => void ) |
    // ( (parent: State, value: Field) => Field )
    (parent: State, value: Field) => ( $NonMaybeType<Field> | void)
  ),
  reflector: Reflector<Field, State>
): {
  reflector: Reflector<Field, State>,
  parent: State,
} {
  const value = reflector.getter(parent)
  const nextValueRaw = updater(parent, value)
  let nextValue: $NonMaybeType<Field>
  if (nextValueRaw === undefined)
    nextValue = reflector.defaultValue
  else
    nextValue = nextValueRaw
  if (nextValue === value) return {
    reflector,
    parent,
  }
  const nextState = reflector.setter(parent, nextValue)
  const result = new Reflector
  result.id = reflector.id
  result.seq = reflector.seq + 1
  result.value = nextValue
  return {
    reflector: result,
    parent: nextState,
  }
}
