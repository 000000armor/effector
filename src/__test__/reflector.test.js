//@flow

import {init, update} from '../reflector'

test('init', () => {
  expect(init(
    {field: 0},
    ({target}) => target,
    (parent, target) => ({
      ...parent,
      target,
    })
  )).toBeDefined()
})

describe('update', () => {
  const state = {
    target: {field: 0},
    another: 'foo',
  }

  const reflector = init(
    {field: 0},
    ({target}) => target,
    (parent, target) => ({
      ...parent,
      target,
    })
  )
  test('should do immutable update', () => {
    expect(update(
      state,
      (parent, {field}) => ({field: field+1}),
      reflector
    )).toMatchSnapshot()
  })
  test('should handle the same and undefined return value', () => {
    expect(update(
      state,
      () => {},
      reflector
    )).toMatchSnapshot()

    expect(update(
      state,
      (parent, value) => value,
      reflector
    ).parent).toBe(state)
  })
})
