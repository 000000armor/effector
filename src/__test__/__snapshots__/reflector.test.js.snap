// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update should do immutable update 1`] = `
Object {
  "parent": Object {
    "another": "foo",
    "target": Object {
      "field": 1,
    },
  },
  "reflector": Reflector {
    "id": 1,
    "seq": 1,
    "value": Object {
      "field": 1,
    },
  },
}
`;

exports[`update should handle the same and undefined return value 1`] = `
Object {
  "parent": Object {
    "another": "foo",
    "target": Object {
      "field": 0,
    },
  },
  "reflector": Reflector {
    "id": 1,
    "seq": 1,
    "value": Object {
      "field": 0,
    },
  },
}
`;
